components/brave_wallet/browser/zcash/zcash_serializer.cc

```diff
@@ -0,0 +1,296 @@
+
+#include <string>
+#include <string_view>
+
+#include "base/big_endian.h"
+#include "base/sys_byteorder.h"
+#include "brave/components/brave_wallet/common/btc_like_serializer_stream.h"
+#include "brave/third_party/argon2/src/src/blake2/blake2.h"
+
+namespace brave_wallet {
+
+namespace {
+
+constexpr char kTransparentHashPersonalizer[] = "ZTxIdTranspaHash";
+constexpr char kSaplingHashPersonalizer[] = "ZTxIdSaplingHash";
+constexpr char kOrchardHashPersonalizer[] = "ZTxIdOrchardHash";
+
+constexpr uint32_t kV5TxVersion = 5 | 1 << 31 /* overwintered bit */;
+constexpr uint32_t kV5VersionGroupId = 0x26A7270A;
+constexpr uint32_t kConsensusBranchId = 0xC2D6D0B4;
+
+std::string GetTxHashPersonalizer() {
+  uint32_t consensus = base::ByteSwapToLE32(kConsensusBranchId);
+  std::string personalizer("ZcashTxHash_");
+  personalizer.append(reinterpret_cast<const char*>(&consensus),
+                      sizeof(kConsensusBranchId));
+  return personalizer;
+}
+
+std::array<uint8_t, 32> blake2b256(const std::vector<uint8_t>& payload,
+                                   std::string_view personalizer) {
+  blake2b_state blake_state;
+  blake2b_param params;
+
+  if (personalizer.length() != sizeof(params.personal)) {
+    NOTREACHED();
+    return {};
+  }
+
+  params.digest_length = 32;
+  params.key_length = 0;
+  params.fanout = 1;
+  params.depth = 1;
+  params.leaf_length = 0;
+  params.node_offset = 0;
+  params.node_depth = 0;
+  params.inner_length = 0;
+  memset(params.reserved, 0, sizeof(params.reserved));
+  memset(params.salt, 0, sizeof(params.salt));
+  memcpy(params.personal, personalizer.data(), sizeof(params.personal));
+  if (blake2b_init_param(&blake_state, &params) != 0) {
+    NOTREACHED();
+    return {};
+  }
+  if (blake2b_update(&blake_state, payload.data(), payload.size()) != 0) {
+    NOTREACHED();
+    return {};
+  }
+  std::array<uint8_t, 32> result;
+  if (blake2b_final(&blake_state, result.data(), 32) != 0) {
+    NOTREACHED();
+    return {};
+  }
+
+  return result;
+}
+
+void PushHeader(const ZCashTransaction& tx, BtcLikeSerializerStream& stream) {
+  stream.Push32AsLE(kV5TxVersion);
+  stream.Push32AsLE(kV5VersionGroupId);
+  stream.Push32AsLE(kConsensusBranchId);
+  stream.Push32AsLE(tx.locktime());
+  stream.Push32AsLE(tx.expiry_height());
+}
+
+void PushOutpoint(const ZCashTransaction::Outpoint& outpoint,
+                  BtcLikeSerializerStream& stream) {
+  stream.PushBytes(outpoint.txid);
+  stream.Push32AsLE(outpoint.index);
+}
+
+void PushOutput(const ZCashTransaction::TxOutput& output,
+                BtcLikeSerializerStream& stream) {
+  stream.Push64AsLE(output.amount);
+  stream.PushSizeAndBytes(output.script_pubkey);
+}
+
+std::array<uint8_t, 32> HashAmounts(const ZCashTransaction& tx) {
+  std::vector<uint8_t> data;
+  BtcLikeSerializerStream stream(&data);
+  for (const auto& input : tx.inputs()) {
+    stream.Push64AsLE(input.utxo_value);
+  }
+  return blake2b256(data, "ZTxTrAmountsHash");
```

Comment:
where's the constants for these blake2b256 being defined? I presume their acting as "nothing up our sleeves" keys and the goal is to just produce the hash value, but I just want to double check my assumptions here.

Reply:
They are defined in https://zips.z.cash/zip-0244

Reply:
makes sense after seeing similar fields in the orchard lib. Do you mind adding a comment to this link?